### ISSUE TITLE
moveToIndex fails, followed by InternalError or segfault

### ISSUE THREAD

**Issue by ollierob** [OpenHFT/Chronicle-Queue#446]
We have a set up where several (~4) threads are writing into many `BINARY_LIGHT`/rolling `HOURLY` queues (30+) which are read by many (<=64) threads in a pool when they are signalled by the writers. Only one of the reader threads can operate on a given queue at any one time.

I have some logic to continue reading from a particular Chronicle index (to "carry on" working after e.g. a program restart or a network disconnect, given a persisted index):
```
tailer.direction(TailerDirection.FORWARD);
if (!tailer.moveToIndex(indexInclusive)) {
    logger.warn("Could not read from index {}! Will attempt to catch up ...", indexInclusive);
    try (final DocumentContext documentContext = tailer.readingDocument()) {
        final long nextIndex = documentContext.index();
        logger.info("Skipped from {} -> {}", indexInclusive, nextIndex);
    }
}
```
Usually this works with multiple threads reading from multiple queues. But on rare occasions - not sure what the trigger for this is yet - they all suddenly start moving to negative indexes, in particular what looks like `Long.MIN_VALUE`:
```
Could not read from index 1815272222624082! Will attempt to catch up ...
Skipped from 1815272222624082 -> -9223372036854775808
```
When this happens I immediately start seeing errors like this (caused by other threads writing to the same queue):
```
Caused by: java.lang.InternalError: a fault occurred in a recent unsafe memory access operation in compiled Java code
        at net.openhft.chronicle.core.UnsafeMemory.readVolatileInt(UnsafeMemory.java:464)
        at net.openhft.chronicle.bytes.NativeBytesStore.readVolatileInt(NativeBytesStore.java:354)
        at net.openhft.chronicle.bytes.MappedBytes.readVolatileInt(MappedBytes.java:533)
        at net.openhft.chronicle.wire.AbstractWire.readFirstHeader(AbstractWire.java:277)
        at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue$StoreSupplier.acquire(SingleChronicleQueue.java:763)
        at net.openhft.chronicle.queue.impl.WireStorePool.acquire(WireStorePool.java:96)
        at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreAppender.setCycle2(SingleChronicleQueueExcerpts.java:244)
        at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreAppender.writingDocument(SingleChronicleQueueExcerpts.java:325)
        at net.openhft.chronicle.wire.MarshallableOut.writingDocument(MarshallableOut.java:57)
```
Or just as frequently JVM crashes in varying locations:
```
INFO   | jvm 5    | 2018/03/21 15:08:04 | # A fatal error has been detected by the Java Runtime Environment:
INFO   | jvm 5    | 2018/03/21 15:08:04 | #
INFO   | jvm 5    | 2018/03/21 15:08:04 | #  SIGBUS (0x7) at pc=0x00007ffa82bc7e6a, pid=25741, tid=0x00007ffa64daa700
INFO   | jvm 5    | 2018/03/21 15:08:04 | #
INFO   | jvm 5    | 2018/03/21 15:08:04 | # JRE version: Java(TM) SE Runtime Environment (8.0_152-b16) (build 1.8.0_152-b16)
INFO   | jvm 5    | 2018/03/21 15:08:04 | # Java VM: Java HotSpot(TM) 64-Bit Server VM (25.152-b16 mixed mode linux-amd64 compressed oops)
INFO   | jvm 5    | 2018/03/21 15:08:04 | # Problematic frame:
INFO   | jvm 5    | 2018/03/21 15:08:04 | # V  [libjvm.so+0xaa1e6a]  Unsafe_CompareAndSwapInt+0xda
```
Deleting the queue off disk seems to fix the problem for a time, until it reoccurs ~hours later. I am unable to reproduce this in other environments, or share much more of the code, so I appreciate that this isn't something that's easily looked into. That being said I would greatly appreciate any insight into what may be causing this (especially if I am doing something wrong in the first code segment!) and how best to recover from it, other than deleting files off disk.

Note that this is using `chronicle-queue:4.6.77` on JVM `8.0_152` running off CentOS:
```
$ cat /proc/version
Linux version 2.6.32-696.20.1.el6.x86_64 (mockbuild@x86_64_01.bsys.centos.org) (gcc version 4.4.7 20120313 (Red Hat 4.4.7-18) (GCC) ) #1 SMP Fri Jan 26 17:51:45 UTC 2018
```

**Comment by ollierob** [comment]
It looks like some other segfault, most likely caused by Chronicle, typically precedes the above scenario:
```
INFO   | jvm 1    | 2018/03/22 03:59:21 | # A fatal error has been detected by the Java Runtime Environment:
INFO   | jvm 1    | 2018/03/22 03:59:21 | #
INFO   | jvm 1    | 2018/03/22 03:59:21 | #  SIGBUS (0x7) at pc=0x00007f82850520ba, pid=6656, tid=0x00007f827d4bf700
INFO   | jvm 1    | 2018/03/22 03:59:21 | #
INFO   | jvm 1    | 2018/03/22 03:59:21 | # JRE version: Java(TM) SE Runtime Environment (8.0_152-b16) (build 1.8.0_152-b16)
INFO   | jvm 1    | 2018/03/22 03:59:21 | # Java VM: Java HotSpot(TM) 64-Bit Server VM (25.152-b16 mixed mode linux-amd64 compressed oops)
INFO   | jvm 1    | 2018/03/22 03:59:21 | # Problematic frame:
INFO   | jvm 1    | 2018/03/22 03:59:21 | # v  ~StubRoutines::jbyte_disjoint_arraycopy
```
Process was automatically restarted and segfaulted again round about the time we start opening queues:
```
INFO   | jvm 2    | 2018/03/22 03:59:53 | # A fatal error has been detected by the Java Runtime Environment:
INFO   | jvm 2    | 2018/03/22 03:59:53 | #
INFO   | jvm 2    | 2018/03/22 03:59:53 | #  SIGBUS (0x7) at pc=0x00007f30d4de3d14, pid=11022, tid=0x00007f3098a4d700
INFO   | jvm 2    | 2018/03/22 03:59:53 | #
INFO   | jvm 2    | 2018/03/22 03:59:53 | # JRE version: Java(TM) SE Runtime Environment (8.0_152-b16) (build 1.8.0_152-b16)
INFO   | jvm 2    | 2018/03/22 03:59:53 | # Java VM: Java HotSpot(TM) 64-Bit Server VM (25.152-b16 mixed mode linux-amd64 compressed oops)
INFO   | jvm 2    | 2018/03/22 03:59:53 | # Problematic frame:
INFO   | jvm 2    | 2018/03/22 03:59:53 | # C  [libc.so.6+0x89d14][thread 139846151022336 also had an error]
INFO   | jvm 2    | 2018/03/22 03:59:53 |   _wordcopy_fwd_dest_aligned+0xa4
```
Then the previously described scenario was observed.

I suspect this may be caused by running out of disk space.

**Comment by dpisklov** [comment]
Are you sharing queue appenders between threads? That might cause this.

**Comment by ollierob** [comment]
I think it's possible for multiple threads to try to append to the same queue, but I am not caching them in any way: they always acquire via `SingleChronicleQueue::acquireAppender`.

**Comment by dpisklov** [comment]
Using acquireAppender is safe as it returns ThreadLocal. As long as you
don't share that is fine.

I noticed you're using ancient version of Linux, I think there was some
issue related to that - @RobAustin @JerryShea do you remember any details
by a chance?

**Comment by RobAustin** [comment]
@ollierob thanks for the code in this issue, could you create a failing test and submit it via a pull request ?

**Comment by ollierob** [comment]
@RobAustin as I mentioned that code usually works, but sometimes under load it triggers a segfault, likely because of other upstream segfaults. It's not easily reproducible, but the following seems to be the lifecycle: 
* Being low on (but not out of) disk space (something like <2gb whilst maintaining 100+ queues) causes a segfault at `~StubRoutines::jbyte_disjoint_arraycopy` for an unknown reason (the stack traces are not useful here to dig further)
* Restarting the JVM and reading from any of the queues previously open has `DocumentContext::index` return a negative index.
* This starts throwing `InternalError` and triggers another segfault as described above.

Since yesterday I have maintained at least 6gb free on disk by deleting the oldest `cq4` files not in use (process is generating about 50gb/day) and the error has not repeated.

Does Chronicle perform any free-space monitoring/assertions or does it delegate this entirely to the OS? (Yes we are running an old kernel but that's CentOS 6 for you.) Is there more it _could_ do in this space?

**Comment by RobAustin** [comment]
The LONG.MIN_VALUE denotes the index is not set. So it’s failing to move to the index. If you had a support contract you could email us the corrupt .cq4  queue file so that we could investigate it for you, if this is an approach that you wish to consider please contact sales@chronicle.software. I’m sorry but It’s goung to be hard to investigate this issue without taking a look at the corrupt queue file. I’ll leave the issue open for now, but it’s likely we are not going to be able to fix this without having the file.

**Comment by RobAustin** [comment]
It sounds like this occurs when you are resource constraint. I’m wondering if your queue has gone into store recovery, this can occur if a write to the queue stalls, but I think the timeout is about 20 seconds. If this occurs I don’t know why this would corrupt your data but it would be worth ruling out.  If this was the case - If you look at the end of the queue you will see an 8Mb empty meta msg starting with a string mentioning store recovery.

**Comment by dpisklov** [comment]
Which roll cycle you use? Given you're running out of space, could it be
that you're experiencing index numeric overflow? Each roll cycle supports
up to certain number of entries per queue file.

**Comment by ollierob** [comment]
Thanks for the suggestions, I will look into them further when I'm back from holiday next week.

**Comment by ollierob** [comment]
My original segfaults were caused by running out of disk space. I can see there is some checking towards this in Chronicle but perhaps there is some room for improvement in this space. Since improving our disk-tidying infrastructure to always maintain at least 10gb free the problem has not reoccured.

In terms of my original move-to-index code snippet, this was also faulty - if I tell Chronicle to move to an arbitrary index it doesn't know about then I cannot expect it to find the nearest-best. Indeed in this case the tailer was never properly initialized and so the document context was not present (hence the `Long.MIN_VALUE` index it returned). We've since replaced this with code that starts from the end of the tailer and moves backwards, until it finds the next index greater-than the target index.
